### PR TITLE
fix(stark-ui): ui dropdown: not taking into account option - emitEvent

### PR DESCRIPTION
**Pull request by nicanac**
ISSUES CLOSED: #1364

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/NationalBankBelgium/stark/blob/master/CONTRIBUTING.md#-commit-message-guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[X] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Change the dropdown setDisabledState function to take into account the emitEvent.

Issue Number: #1364 

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

## Other information

### CODE REVIEW

File: packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts [hunk 1]
```diff
@@ -561,6 +561,47 @@ describe("DropdownComponent", () => {
 				});
 			});
 
+			describe("DropDown properties", () => {
```
Comment (SuperITMan) [inline packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts]: Could you please change the description for something more related to what you will test ? 😊

File: packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts [hunk 2]
```diff
@@ -561,6 +561,47 @@ describe("DropdownComponent", () => {
...
+				let mockObserver: SpyObj<Observer<any>>;
+
+				beforeEach(() => {
+					mockObserver = createSpyObj<Observer<any>>("observerSpy", ["next", "error", "complete"]);
+					component = hostComponent.dropdownComponent;
```
Comment (SuperITMan) [inline packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts]: This line is not needed since it's already done in the `beforeEach` at line 500

File: packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts [hunk 3]
```diff
@@ -561,6 +561,47 @@ describe("DropdownComponent", () => {
...
+				});
+
+				it("the MatOption should be disabled when the form control is disabled AND it should trigger a 'valueChange' event ONLY IF the 'emitEvent' option is enabled", () => {
```
Comment (SuperITMan) [inline packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts]: Could you please split this test in 2 parts (one with `emitEvent: false`, the other with `emitEvent: true` ?
Also, could you rephrase the `it` ?  
What you want to test here is that when you disable the `formControl` with(out) `emitEvent`, it reacts as expected. 
You don't care at this moment if the MatOption is disabled or not. Actually this is checked already at line 552 😊

File: packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts [hunk 4]
```diff
@@ -561,6 +561,46 @@ describe("DropdownComponent", () => {
...
+				});
+
+				it("When the formControl is disabled or enabled with emitEvent with emitEvent set to true the dropdown shouldn't trigger a 'valueChange' event", () => {
+					hostComponent.formControl.valueChanges.subscribe(mockObserver);
+
+					hostComponent.formControl.disable(); // 'emitEvent' true by default
+					hostFixture.detectChanges();
+
+					expect(component.disabled).toBe(true);
+					expect(mockObserver.next).toHaveBeenCalledTimes(1);
```
Comment (SuperITMan) [inline packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts]: Could you please add those 2 checks ?
```typescript
expect(mockObserver.error).not.toHaveBeenCalled();
expect(mockObserver.complete).not.toHaveBeenCalled();
```

Otherwise, if they are called after (line 599), you cannot be sure they were not called before...

Comment (nicanac) [reply packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts]: added

File: packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts [hunk 5]
```diff
@@ -561,6 +561,46 @@ describe("DropdownComponent", () => {
 				});
 			});
 
+			describe("Changing emitEvent properties ", () => {
```
Comment (SuperITMan) [inline packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts]: This `describe` is not really relevant since you are not "changing emitEvent properties". Actually, you are disabling/enabling the dropdown with or without the `emitEvent; false`.
Maybe you could move your 2 `it` into `describe("formControl.disabled", ()`. What do you think @nicanac ?

In that case, I would advise to you to declare the mockObserver in each `it` since it's not necessary for 
```
it("should set the right value to the mat-select when 'formControl.disabled' changes", () => {
```

Comment (nicanac) [reply packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts]: yes, why not, seems legit for me.

Comment (nicanac) [reply packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts]: moved to the `describe("formControl.disabled", ().`

File: packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts [hunk 6]
```diff
@@ -559,6 +559,48 @@ describe("DropdownComponent", () => {
 					hostFixture.detectChanges();
 					expect(selectElement.outerHTML).toMatch(/ng-reflect-disabled="false"/);
 				});
+
+				it("Shouldn't trigger a 'valueChange' event when the formControl is disabled or enabled with emitEvent set to false", () => {
```
Comment (SuperITMan) [inline packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts]: could you please change this in lowercase as it is the case in the other tests ? 😊 
```
it("shouldn't trigger a 'valueChange' event when the formControl is di ...
```

File: packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts [hunk 7]
```diff
@@ -559,6 +559,48 @@ describe("DropdownComponent", () => {
...
+					hostComponent.formControl.enable({ emitEvent: false });
+					hostFixture.detectChanges();
+
+					expect(component.disabled).toBe(false);
+					expect(mockObserver.next).not.toHaveBeenCalled(); // because the 'emitEvent' is false
+					expect(mockObserver.error).not.toHaveBeenCalled();
+					expect(mockObserver.complete).not.toHaveBeenCalled();
+				});
+
+				it("Shouldn't trigger a 'valueChange' event when the formControl is disabled or enabled with emitEvent set to true", () => {
```
Comment (SuperITMan) [inline packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts]: I think that it "should" trigger in this case :wink:

Comment (christophercr) [inline packages/stark-ui/src/modules/dropdown/components/dropdown.component.spec.ts]: "should trigger ..."
